fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/MacroFunction.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * MacroFunction.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2025 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.query.plan.cascades;
+
+import com.apple.foundationdb.record.PlanHashable;
+import com.apple.foundationdb.record.PlanSerializationContext;
+import com.apple.foundationdb.record.RecordMetaDataProto;
+import com.apple.foundationdb.record.metadata.SerializableFunction;
+import com.apple.foundationdb.record.planprotos.PMacroFunctionValue;
+import com.apple.foundationdb.record.query.plan.cascades.typing.Type;
+import com.apple.foundationdb.record.query.plan.cascades.typing.Typed;
+import com.apple.foundationdb.record.query.plan.cascades.values.QuantifiedObjectValue;
+import com.apple.foundationdb.record.query.plan.cascades.values.Value;
+import com.apple.foundationdb.record.query.plan.cascades.values.translation.TranslationMap;
+import com.apple.foundationdb.record.query.plan.serialization.DefaultPlanSerializationRegistry;
+
+import javax.annotation.Nonnull;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * MacroFunction that expands a body (referring to parameters) into a {@link Value} (through encapsulation) call site.
+ */
+public class MacroFunction implements SerializableFunction {
+    @Nonnull
+    private final String functionName;
```

Comment:
All functions will have this in common:
1. function name
2.declared parameters (types at least)
3.return type 
Can you move that up into `SerializableFunction`? You can give `SerializableFunction` the appropriate getters forcing all implementors to provide methods for them. As a second step, have an abstract base class implementing `SerializableFunction`. Let's call that `AbstractCatalogedFunction`. All your methods extends from that. It's the same approach `Value`, and `AbstractValue` (for instance) have taken.

Reply:
done, restructured the code.